### PR TITLE
expression: support NO_ZERO_DATE sql_mode

### LINKED ISSUE #11403: Insert zero date successfully in default sql mode
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
create table t(a date);
insert into t values('0000-00-0');
```

Relate issue: #8637

2. What did you expect to see?
```
mysql> create table t(a date);
Query OK, 0 rows affected (0.07 sec)

mysql> insert into t values('0000-00-0');
ERROR 1292 (22007): Incorrect date value: '0000-00-0' for column 'a' at row 1
```

3. What did you see instead?
```
mysql> create table t(a date);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t values('0000-00-0');
Query OK, 1 row affected (0.00 sec)

mysql> select * from t;
+------------+
| a          |
+------------+
| 0000-00-00 |
+------------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
### What problem does this PR solve?

Issue Number: close #11403

### What is changed and how it works?

Check the `NO_ZERO_DATE` sql_mode for date and datetime.
This PR will fix the behavior of inserting and `str_to_date`, however,  `cast(xx as date)` isn't fixed by this PR because it's too complex to do it.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Breaking backward compatibility

### Release note 
support NO_ZERO_DATE sql_mode.

### CODE REVIEW

File: table/column.go [hunk 1]
```diff
@@ -190,6 +190,10 @@ func CastValue(ctx sessionctx.Context, val types.Datum, col *model.ColumnInfo) (
 		truncateTrailingSpaces(&casted)
 	}
 
+	if ((col.Tp == mysql.TypeDate && !casted.IsNull() && casted.GetMysqlTime() == types.ZeroDate) || (col.Tp == mysql.TypeDatetime && !casted.IsNull() && casted.GetMysqlTime() == types.ZeroDatetime)) && ctx.GetSessionVars().SQLMode.HasNoZeroDateMode() {
```
Comment (imtbkcat) [inline table/column.go:193]: MySQL doc said:

If this mode is enabled, '0000-00-00' is permitted and inserts produce a warning.
If this mode and strict mode are enabled, '0000-00-00' is not permitted and inserts produce an error, unless IGNORE is given as well. For INSERT IGNORE and UPDATE IGNORE, '0000-00-00' is permitted and inserts produce a warning.

So we should check `STRICT_TRANS_MODE`

Comment (wjhuang2016) [reply table/column.go:193]: It's checked in handleInvalidTimeError

**Comment by wjhuang2016** [comment]
> `vecEvalDuration` also should be modified.

Addressed, PTAL.

**Comment by zz-jason** [comment]
@wjhuang2016 integration test failed. PTAL.

**Pull request merged**